Technical Advisory 104309 (#18309)
cockroachdb/docs: src/current/advisories/a104309.md

---
title: Technical Advisory a104309
advisory: A-a104309

> summary: In rare cases, a rangefeed bug may cause a checkpoint to be emitted prematurely in rare cases, before all writes below the checkpoint timestamp have been emitted.

Reviewer: nit: "in rare cases" is written twice

> summary: In rare cases, a rangefeed bug may cause a checkpoint to be emitted prematurely in rare cases, before all writes below the checkpoint timestamp have been emitted.

Author: ```suggestion
summary: In rare cases, a rangefeed bug may cause a checkpoint to be emitted prematurely, before all writes below the checkpoint timestamp have been emitted.
```

toc: true

> affected_versions: v1.0 to v22.2.17, v23.1.0 to v23.1.14, v23.2.0

Reviewer: nit: did we have changefeeds in 1.0? Doesn't matter for practical purposes.

Author: Good call, they seem to have appeared in 2.1.11.

> affected_versions: v1.0 to v22.2.17, v23.1.0 to v23.1.14, v23.2.0

Author: ```suggestion
affected_versions: v2.1.11 to v22.2.17, v23.1.0 to v23.1.14, v23.2.0
```

1. If the above conditions are satisfied, the follower attempts to push the transaction to advance its resolved timestamp and emit a checkpoint, but the follower does not find a transaction record. The follower operates as though the transaction was aborted, and allows its resolved timestamp to advance above the transaction's write timestamp, emitting a checkpoint even though the intents have not yet been emitted.

1. When the intent resolution is finally replicated to the follower, it emits the write events with timestamps below the previous checkpoint.

1. The changefeed processor detects these events below the previous checkpoint and discards them instead of emitting them.

All CockroachDB versions prior to and including the following are impacted by this bug:

> - v22.2.17 and below

Author: ```suggestion
- v2.1.11 - v22.2.17
```

The fix has been applied to maintenance releases of CockroachDB:

- [v22.2.18](https://www.cockroachlabs.com/docs/releases/v22.2#v22-2-18)
- [v23.1.15](https://www.cockroachlabs.com/docs/releases/v23.1#v23-1-15)
- [v23.2.1](https://www.cockroachlabs.com/docs/releases/v23.2#v23-2-1)

This public issue is tracked by [Issue #104309](https://github.com/cockroachdb/cockroach/issues/104309).

## Mitigation

Users are encouraged to upgrade to v22.2.18, v23.1.15, v23.2.1, or a later version that includes the fix.

The log message `cdc ux violation: detected timestamp ... that is less or equal to the local frontier` indicates that a cluster has been affected by this bug. One message will be logged per omitted event. The message will include the [job ID](https://www.cockroachlabs.com/docs/v23.2/show-jobs#show-changefeed-jobs) of the affected changefeed (as `job=...`), as well as the MVCC timestamp of the transaction whose writes were omitted (`detected timestamp ...`). If the data is still present in a table, the following query may allow you to retrieve it:

{% include_cached copy-clipboard.html %}
~~~ sql

> SELECT * FROM table WHERE crdb_internal_mvcc_timestamp = {timestamp}

Reviewer: We'll need this qualifier:

Where `timestamp` must be converted to nanoseconds with the logical component after the decimal point. For example, given the logged timestamp 1707854079.38430848,2 multiply 1707854079 by 1e9 (1 billion) and add 38430848, then put 2 after the decimal point: 1707854079038430848.2

Author: Done, with some light tweaking.

~~~

> 

Author: ```suggestion
Replace '{timestamp}` with the timestamp expressed in nanoseconds with the logical component after the decimal point. For example, to query for the logged timestamp `1707854079.38430848,2`:

1. Multiply `1707854079` by <code>1 × 10<sup>9</sup> (1 billion).
1. Add `38430848` to the result.
1. Append the portion of the timestamp (`2` in this example) after the comma after the decimal point. The converted timestamp is  `1707854079038430848.2`.
```

{% include_cached copy-clipboard.html %}
~~~ sql
SELECT * FROM table WHERE crdb_internal_mvcc_timestamp = {timestamp}
~~~
Replace '{timestamp}` with the timestamp expressed in nanoseconds with the logical component after the decimal point. For example, to query for the logged timestamp `1707854079.38430848,2`:

1. Multiply `1707854079` by <code>1 × 10<sup>9</sup></code> (1 billion).
1. Add `38430848` to the result.
1. Append the portion of the timestamp (`2` in this example) after the comma after the decimal point. The converted timestamp is `1707854079038430848.2`.
If data is found, it can be re-emitted in two ways:

- By issuing an `UPDATE` command with the same values.

> - By restarting the changefeed with `initial_scan` set to `yes` and without a cursor. This will perform a full export of all current data across the changefeed. If a cursor is provided, an initial scan will not be performed.

Reviewer 2: Is there a way to initialize a cursor with the timestamp of missing writes ?

The log message `cdc ux violation: detected timestamp ... that is less or equal to the local frontier` indicates that a cluster has been affected by this bug. One message will be logged per omitted event. The message will include the [job ID](https://www.cockroachlabs.com/docs/v23.2/show-jobs#show-changefeed-jobs) of the affected changefeed (as `job=...`), as well as the MVCC timestamp of the transaction whose writes were omitted (`detected timestamp ...`). If the data is still present in a table, the following query may allow you to retrieve it:

{% include_cached copy-clipboard.html %}
~~~ sql
SELECT * FROM table WHERE crdb_internal_mvcc_timestamp = {timestamp}
~~~
Replace '{timestamp}` with the timestamp expressed in nanoseconds with the logical component after the decimal point. For example, to query for the logged timestamp `1707854079.38430848,2`:

1. Multiply `1707854079` by <code>1 × 10<sup>9</sup></code> (1 billion).
1. Add `38430848` to the result.
1. Append the portion of the timestamp (`2` in this example) after the comma to the right of the the decimal point. In this example, the converted timestamp is `1707854079038430848.2`.

> If data is found, it can be re-emitted in two ways:

Author: ```suggestion
If data is found, it can be re-emitted in any of the following ways:
```

---
title: Technical Advisory a104309
advisory: A-a104309
summary: In rare cases, a rangefeed bug may cause a checkpoint to be emitted prematurely, before all writes below the checkpoint timestamp have been emitted.
toc: true
affected_versions: v2.1.11 to v22.2.17, v23.1.0 to v23.1.14, v23.2.0

> advisory_date: 2024-02-20

Author: ```suggestion
advisory_date: 2024-02-21
```

docs_area: releases
---

Publication date: {{ page.advisory_date | date: "%B %e, %Y" }}

## Description

> A [rangefeed](https://www.cockroachlabs.com/docs/stable/create-and-configure-changefeeds#enable-rangefeeds) bug may cause a [checkpoint](https://www.cockroachlabs.com/docs/stable/how-does-an-enterprise-changefeed-work) to be emitted prematurely in rare cases, before all writes below the checkpoint timestamp have been emitted. This could occur if nodes are overloaded or if long-running transactions (several seconds) are involved, such that intent resolution only replicates to a follower running a rangefeed more than 10 seconds after the transaction began. If a cluster using changefeeds experiences this bug, changefeeds will omit these write events, and the following error will be logged:

Reviewer 3: We should link the expression, "write intent" to some reference reading. I think this is probably the best spot. https://www.cockroachlabs.com/docs/v23.2/architecture/transaction-layer#write-intents

Author: I can't see any mention of "write intents" but "write events". I added the link to "intent resolution" instead. WDYT?

> A [rangefeed](https://www.cockroachlabs.com/docs/stable/create-and-configure-changefeeds#enable-rangefeeds) bug may cause a [checkpoint](https://www.cockroachlabs.com/docs/stable/how-does-an-enterprise-changefeed-work) to be emitted prematurely in rare cases, before all writes below the checkpoint timestamp have been emitted. This could occur if nodes are overloaded or if long-running transactions (several seconds) are involved, such that intent resolution only replicates to a follower running a rangefeed more than 10 seconds after the transaction began. If a cluster using changefeeds experiences this bug, changefeeds will omit these write events, and the following error will be logged:

Author: ```suggestion
A [rangefeed](https://www.cockroachlabs.com/docs/stable/create-and-configure-changefeeds#enable-rangefeeds) bug may cause a [checkpoint](https://www.cockroachlabs.com/docs/stable/how-does-an-enterprise-changefeed-work) to be emitted prematurely in rare cases, before all writes below the checkpoint timestamp have been emitted. This could occur if nodes are overloaded or if long-running transactions (several seconds) are involved, such that [intent resolution](https://www.cockroachlabs.com/docs/stable/architecture/transaction-layer#write-intents) only replicates to a follower running a rangefeed more than 10 seconds after the transaction began. If a cluster using changefeeds experiences this bug, changefeeds will omit these write events, and the following error will be logged:
```

{% include_cached copy-clipboard.html %}
~~~ none
cdc ux violation: detected timestamp ... that is less or equal to the local frontier
~~~

Across a large sample of clusters using changefeeds over a month, 3.8% of clusters had experienced this bug, and a total 0.000064% of events were omitted (one in a million).

The bug is caused by a race condition that is typically triggered by transactions that run for several seconds or by follower replicas that lag significantly behind the leaseholder, such as when a node is overloaded or a network is disrupted. The rangefeed must be running on a follower replica that applies intent resolution long after the original transaction has committed and garbage-collected its intents and transaction record on the leaseholders (at least 10 seconds after the transaction begins).

The specific conditions to trigger the bug are:

> 1. An explicit or cross-range transaction commits, then asynchronously resolves all of its intents and removes its transaction record on all relevant leaseholders.

Reviewer 3: Explicit OR cross-range transactions is a strange subset of all transactions. Is it explicit AND cross-range? 

Also, aren't all write intents removed async after the transaction commits during the cleanup stage? The language here just doesn't make sense to me.

Reviewer 3: The way this is written reads (to me) as if this behavior is unusual.

Reviewer 2: This might be worth clarifying. The intention here is "All explicit transactions and cross-range implicit transactions are affected."

Reviewer 2: The reason is that single-range implicit transactions don't write intents.

Author: @shralex Is the suggested edit accurate? https://github.com/cockroachdb/docs/pull/18309/files#r1496690828 (sorry about the HTML, it's a tooling edge case)

Reviewer 2: yes, LGTM!

> 1. An explicit or cross-range transaction commits, then asynchronously resolves all of its intents and removes its transaction record on all relevant leaseholders.

Author: ```suggestion
1. One of the following types of transactions commits,  then asynchronously resolves all of its intents and removes its transaction record on all relevant leaseholders.<ul><li>An <a href="https://www.cockroachlabs.com/docs/stable/transactions#sql-statements">explicit transaction</a>.</li><li>An implicit transaction (single-statement write) that crosses ranges.</li></ul>
```

Author: @shannonbradshaw Is this more clear?

> 1. A follower replica running a rangefeed replicates the transaction's intent writes, but not yet the resolution of the intents. More than 10 seconds must elapse between when the transaction begins and when the intent resolution is replicated to the follower. (The intent resolution must be replicated to the follower after 10 seconds. That means that the sum of `transaction duration + intent resolution duration + transaction record GC + replication to follower` must be greater than 10 seconds to trigger the bug.)

Reviewer 3: This sentence seem unnecessary, if not redundant -- "The intent resolution must be replicated to the follower after 10 seconds."

> 1. A follower replica running a rangefeed replicates the transaction's intent writes, but not yet the resolution of the intents. More than 10 seconds must elapse between when the transaction begins and when the intent resolution is replicated to the follower. (The intent resolution must be replicated to the follower after 10 seconds. That means that the sum of `transaction duration + intent resolution duration + transaction record GC + replication to follower` must be greater than 10 seconds to trigger the bug.)

Author: ```suggestion
1. A follower replica running a rangefeed replicates the transaction's intent writes, but not yet the resolution of the intents. More than 10 seconds must elapse between when the transaction begins and when the intent resolution is replicated to the follower. (The sum of `transaction duration + intent resolution duration + transaction record GC + replication to follower` must be greater than 10 seconds to trigger the bug.)
```
